framework/SupervisedLearning/ARMA.py

```diff
@@ -1102,6 +1152,120 @@ def finalizeGlobalRomSegmentEvaluation(self, settings, evaluation, weights=None)
         evaluation[target] = self._transformThroughInputCDF(evaluation[target], dist, weights)
     return evaluation
 
+  ### Peak Picker ###
+  def _peakPicker(self,signal,low):
+    """
+      Peak picker
+      @ In, signal, np.array(float), signal to transform
+      @ In, low, float, required height of peaks.
+      @ Out, peaks, np.array, indices of peaks in x that satisfy all given conditions
+      @ Out, heights, np.array, boolean mask where is the residual signal
+    """
+    peaks, properties = find_peaks(signal, height=low)
+    heights = properties['peak_heights']
+    return peaks,heights
+
+  def rangeWindow(self,windowDict):
+    """
+      Collect the window index in to groups
+      @ In, windowDict, dict, dictionary for specefic target peaks
+      @ Out, rangeWindow, list, list of dictionaries which store the window index for each target
+    """
+    rangeWindow = []
+    windowType = len(windowDict['windows'])
+    windows = windowDict['windows']
+    period = windowDict['period']
+    for i in range(windowType):
+      windowRange={}
+      bg_P=(windows[i]['window'][0]-1)%period
+      end_P=(windows[i]['window'][1]+2)%period
+      timeInd=np.arange(len(self.pivotParameterValues))
+      bg_P_ind  = np.where(timeInd%period==bg_P )[0].tolist()
+      end_P_ind = np.where(timeInd%period==end_P)[0].tolist()
+      if bg_P_ind[0]>end_P_ind[0]:
+        tail=end_P_ind[0]
+        end_P_ind.pop(0)
+        end_P_ind.append(tail)
+      windowRange['bg']=bg_P_ind
+      windowRange['end']=end_P_ind
+      rangeWindow.append(windowRange)
+    return rangeWindow
+
+  def _peakGroupWindow(self,signal,windowDict):
+    """
+      Collect the peak information in to groups, define the residual signal
+      @ In, signal, np.array(float), signal to transform
+      @ In, windowDict, dict, dictionary for specefic target peaks
+      @ Out, groupWin, list, list of dictionaries which store the peak information
+      @ Out, maskRes, np.array, boolean mask where is the residual signal
+    """
+    groupWin = []
+    maskRes = signal == signal
```

Comment:
does `signal` ever `!= signal`?

Is this equivalent to `maskRes = np.ones(len(signal), dtype=bool)`? I don't know what the speed difference is, but I bet it makes a difference.